src/app/clusters/door-lock-server/door-lock-server.cpp

```diff
@@ -553,15 +553,15 @@ void DoorLockServer::setCredentialCommandHandler(
         emberAfDoorLockClusterPrintln("[SetCredential] Credential index is out of range [endpointId=%d,credentialType=%u"
                                       ",credentialIndex=%d]",
                                       commandPath.mEndpointId, to_underlying(credentialType), credentialIndex);
-        sendSetCredentialResponse(commandObj, commandPath, DlStatus::kInvalidField, 0, nextAvailableCredentialSlot);
+        sendSetCredentialResponse(commandObj, commandPath, DlStatus::kInvalidField, userIndex.Value(), nextAvailableCredentialSlot);
```

Comment:
I might be missing something, but:

1.  Won't this crash if `userIndex` is null (i.e. this is an Add)?
2. If `userIndex` is not null (i.e. this is a Modify), then per spec it should be null in the response, so we should not pass its value here.

Same for a bunch of the rest of these.

Am I missing something?

Reply:
I think you may have "userIndex" and and "createdUserIndex" conflated. 

The reason for issue 20930 is because 0 is hardcoded as userIndex for all these calls to sendSetCredentialResponse.

The 0 you are thinking of, which indicates to add a user is the createdUserIndex. Poor name in my opinion. You can see in createCredential, it has both a "userIndex" and a "createdUserIndex". User index indicates where in an array the user should be, and createdUserIndex indicates the operation.

Reply:
I agree with #2. Good catch, modify should have null in the response. But I think the rest should be passed the actual userIndex. That seems to be causing issue 20930

Reply:
As far as I can tell from the spec the way this command works is:

1. For a Modify, there is an incoming userIndex.  In this case the result must not have a userIndex.
2. For an Add, there may or may not be an incoming userIndex, depending on whether we are adding a credential to an existing user or adding a credential along with a user.  The spec is extremely unclear about what happens in this case; see https://github.com/CHIP-Specifications/connectedhomeip-spec/issues/5507

Given the endemic lack of responses on the spec bits, let's just implement something sane here and be done with it and the spec can follow.  I'd propose we do the following:

* For any failure, the user index in the response should be null.
* For a Modify, user index in response should be null.
* For an add to an existing user, the user index in the response should be null.
* For an add that creates a new user, the response should indicate the user index of the newly created user.

Reply:
> User index indicates where in an array the user should be, and createdUserIndex indicates the operation.

No, createdUserIndex does not indicate the operation.  It's an outparam that gets set to the index where the user is actually created.

Reply:
Oh I see. I misinterpreted createdUserIndex. operationType determines the operation.

Reply:
I agree with your approach. I didn't consider the case of adding a credential along with a user. With manual testing, I usually test creating a user, then adding a credential to it afterwards.

What I am implementing:
- All failures result as null for userIndex
- For modify, userIndex responds as null
- **Clear user returns userIndex** let me know what you think of this one. Seems sane to me.

The last 2 bullets should be taken care of by changing line 632 to 
        sendSetCredentialResponse(commandObj, commandPath, status, userIndex.Value(), nextAvailableCredentialSlot);

Reply:
> Clear credential returns userIndex let me know what you think of this one. Seems sane to me.

You mean the "Clear Credential" command?  That returns a status response....

Or do you mean something else?

Reply:
Sorry what I meant here is that clear credential should return a **credential** index in the status response rather than null. 

Based on your comment, I see now that clear credential is supposed to be an error.